### PR TITLE
Protect DomainParticipant set_listener [16040]

### PR DESCRIPTION
**Pull request by juanlofer-eprosima**
## Description

Related to #3040 
 This PR modifies the current behaviour of `DomainParitcipant::set_listener`, so that for it to proceed no callbacks should be on execution at that moment. This would prevent the following situation: user sets participant's listener to `nullptr` and destroys the object (listener), while `DomainParticipantImpl::MyRTPSParticipantListener` is about to invoke participant's listener callback (obtained reference was still not `nullptr`), hence resulting in undefined behaviour.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/domain/DomainParticipantImpl.hpp
Comment (MiguelBarro) [inline src/cpp/fastdds/domain/DomainParticipantImpl.hpp:608]: I would leave `if (on_guard_)` alone and assert the value of `listener_->participant_ != nullptr` this way if we introduce inconsistencies they can be readily detected.

File: src/cpp/fastdds/domain/DomainParticipantImpl.hpp
Comment (MiguelBarro) [inline src/cpp/fastdds/domain/DomainParticipantImpl.hpp:614]: why not `notify = !listener_->callback_counter_`?

File: src/cpp/fastdds/domain/DomainParticipantImpl.hpp
Comment (MiguelBarro) [inline src/cpp/fastdds/domain/DomainParticipantImpl.hpp:685]: We should assert `callback_counter_ == 0` on destruction too.

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp
Comment (MiguelBarro) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:1537]: We no longer need local variables. If we have races we want TSAN to report them. We can use `participant_->get_listener()` confidently because the `sentry` is on guard.

**Review by MiguelBarro — CHANGES_REQUESTED** [review]
Some minor comments. Logic is sound👏.

File: src/cpp/fastdds/domain/DomainParticipantImpl.hpp
Comment (juanlofer-eprosima) [reply src/cpp/fastdds/domain/DomainParticipantImpl.hpp:685]: I changed the condition to `!(callback_counter_ > 0)` because that's what we really want to assert. The counter should only be `-1` if listener is `nullptr`, or `0` otherwise.

Comment (MiguelBarro) [reply src/cpp/fastdds/domain/DomainParticipantImpl.hpp:674]: Absolutely right!

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp
Comment (MiguelBarro) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:1537]: I simply say that doing:
```
    Sentry sentinel(this);
    if (sentinel)
    {
        participant_->get_participant()->get_listener()->on_participant_discovery(participant, std::move(info));
    }
```
is perfectly safe because of `Sentry` magic.

Comment (juanlofer-eprosima) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp]: I removed the local variables and added two new conditions to `SEntry` constructor.

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp
Comment (MiguelBarro) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:240]: if it should never happen why the wait? Leave the assert alone 😅

File: test/unittest/dds/participant/ParticipantTests.cpp
Comment (MiguelBarro) [inline test/unittest/dds/participant/ParticipantTests.cpp:2194]: I would have not used an external promise but kept it as a member and return the future via a getter. Nevertheless is a style thing and is up to you.

**Review by MiguelBarro — CHANGES_REQUESTED** [review]
LGTM. Let's wait for the CI.

**Review by MiguelBarro — DISMISSED** [review]
LGTM if it does for TSAN 😎

**Review by MiguelBarro — DISMISSED** [review]
LGTM

**Comment by EduPonz** [comment]
@richiprosima please test this

**Review by jsantiago-eProsima — APPROVED** [review]
LGTM with green CI

**Comment by jsantiago-eProsima** [comment]
@richiprosima please test windows

**Comment by MiguelCompany** [comment]
CI for 2.9.0-rc:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_linux/3187)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_mac/2944)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/fastdds_windows/3349/)
* Discovery server [Build Status](http://jenkins.eprosima.com:8080/job/discovery_server_linux/3131/)

**Pull request merged**
